File: Codigo/backend/app/schemas/knowledge.py
from datetime import datetime

from pydantic import BaseModel


class KnowledgeNodeOut(BaseModel):
    uid: str
    label: str
    text: str
    mention_count: int


class KnowledgeEdgeOut(BaseModel):
    source: str
    target: str
    weight: int


class KnowledgeGraphResponse(BaseModel):
    status: str
    updated_at: datetime | None
    nodes: list[KnowledgeNodeOut]
    edges: list[KnowledgeEdgeOut]


class MentionOut(BaseModel):
    uid: str
    sentence_text: str
    source_type: str  # "document" | "annotation"
    source_uid: str
    source_title: str
    page_number: int | None = None


class MentionsListResponse(BaseModel):
    node: KnowledgeNodeOut
    mentions: list[MentionOut]


class CoOccurrenceOut(BaseModel):
    sentence_text: str
    source_type: str
    source_uid: str
    source_title: str
    page_number: int | None = None


class CoOccurrencesResponse(BaseModel):
    node_a: KnowledgeNodeOut
    node_b: KnowledgeNodeOut
    weight: int
    co_occurrences: list[CoOccurrenceOut]

File: Codigo/mobile/src/types/knowledge.ts
export type EntityLabel = 'PER' | 'LOC' | 'ORG';

export type KnowledgeStatus = 'IDLE' | 'PROCESSING' | 'DONE' | 'FAILED';

export interface KnowledgeNode {
  uid: string;
  label: EntityLabel;
  text: string;
  mention_count: number;
}

export interface KnowledgeEdge {
  source: string;
  target: string;
  weight: number;
}

export interface KnowledgeGraphResponse {
  status: KnowledgeStatus;
  updated_at: string | null;
  nodes: KnowledgeNode[];
  edges: KnowledgeEdge[];
}

export interface Mention {
  uid: string;
  sentence_text: string;
  source_type: 'document' | 'annotation';
  source_uid: string;
  source_title: string;
  page_number: number | null;
}

export interface MentionsListResponse {
  node: KnowledgeNode;
  mentions: Mention[];
}

export interface CoOccurrence {
  sentence_text: string;
  source_type: 'document' | 'annotation';
  source_uid: string;
  source_title: string;
  page_number: number | null;
}

export interface CoOccurrencesResponse {
  node_a: KnowledgeNode;
  node_b: KnowledgeNode;
  weight: number;
  co_occurrences: CoOccurrence[];
}

export type GraphSelection =
  | { type: 'node'; uid: string }
  | { type: 'edge'; a_uid: string; b_uid: string };
